src/main/java/com/peerrequest/app/api/CategoriesController.java

```diff
@@ -24,80 +23,76 @@
  */
 @RestController
 @ApiControllerPrefix
-public class CategoriesController {
-    private Category mockUpCategory;
+public class CategoriesController extends ServiceBasedController {
+    private final int maxPageSize = 100;
 
-    public CategoriesController() {
-        mockCategory();
-    }
+    @GetMapping("/categories")
+    List<Category.Dto> listCategories(@RequestParam Optional<Integer> limit, @RequestParam Optional<Long> after) {
+        if (limit.isPresent()) {
+            if (limit.get() < 0) {
+                throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "limit must be greater than 0");
+            }
+            limit = Optional.of(Math.min(limit.get(), maxPageSize));
+        }
 
-    @GetMapping(value = "/categories", produces = "application/json")
-    public List<Category> getCategories(@AuthenticationPrincipal OAuth2User oauth2User) {
-        return List.of(mockUpCategory);
+        return this.categoryService.list(after.orElse(null), limit.orElse(maxPageSize), null).stream()
+            .map(Category::toDto).toList();
     }
 
-    @GetMapping(value = "/categories/{categoryId}", produces = "application/json")
-    public Category getCategory(@PathVariable final int categoryId) {
-        return mockUpCategory;
-    }
+    @GetMapping("/categories/{id}")
+    Category.Dto getCategory(@PathVariable Long id) {
+        var option = this.categoryService.get(id);
+        if (option.isEmpty()) {
+            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "category does not exist");
+        }
 
-    @PostMapping(value = "/categories", produces = "application/json")
-    public Category createCategory(@RequestBody final Category category) {
-        this.mockUpCategory = category;
-        return mockUpCategory;
+        return option.get().toDto();
     }
 
-    @PatchMapping(value = "/categories/{categoryId}", consumes = "application/json", produces = "application/json")
-    public Category patchCategory(@RequestBody final Category category) {
-        return mockUpCategory;
+    @DeleteMapping("/categories/{id}")
+    Optional<Category.Dto> deleteCategory(@PathVariable Long id) {
```

Comment:
check is missing if the calling user is the creator of the category when the category is external